_overviews/overview/a-taste-of-scala.md

```diff
@@ -0,0 +1,1329 @@
+  def leftTrim(s: String) = s.replaceAll("^\\s+", "")
+
+  // Right-trim a string: `"  foo  "` becomes `"  foo"`.
+  def rightTrim(s: String) = s.replaceAll("\\s+$", "")
+
+  def isNullOrEmpty(s: String): Boolean = 
+    if (s==null || s.trim.equals("")) true else false
+
+end StringUtils
+```
+
+Those functions can be called like this:
+
+```scala
+StringUtils.leftTrim("  hi  ")   // "hi  "
+```
+
+or this:
+
+```scala
+import StringUtils._
+leftTrim("  hi  ")   // "hi  "
+```
+
+
+### Lambdas
+
+Lambdas — also known as *anonymous functions* — are a big part of keeping your code concise but readable. These two examples are equivalent, and show how to multiply each number in a list by `2` by passing a lambda into the `map` method:
+
+```scala
+val a = List(1,2,3).map(i => i * 2)   // List(2,4,6)
+val b = List(1,2,3).map(_ * 2)        // List(2,4,6)
+```
+
+Those examples are also equivalent to that code, but they use a `double` method inside of `map` instead of a lambda:
+
+```scala
+def double(i: Int): Int = i * 2
+val a = List(1,2,3).map(i => double(i))
+val b = List(1,2,3).map(double)
+```
+
+>If you haven’t seen the `map` method before, it applies a given function to every element in a list, yielding a new list that contains the new values.
+
+
+### Higher-order functions
+
+A “higher order function” (HOF) is a function that can take a function as an input parameter, or returns a function as a return value. HOFs are extremely common in Scala, such as being used throughout the standard libraries. These examples show how to pass methods and anonymous functions into the functional methods of Scala collections:
+
+```scala
+// a sample list
+val nums = (1 to 10).toList   // List(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
+
+// sample methods
+def isEven(i: Int): Boolean = i % 2 == 0
+def double(i: Int): Int = i * 2
+
+// pass in a method or function
+val a = nums.filter(isEven)   // List(2, 4, 6, 8, 10)
+val b = nums.map(double)      // List(2, 4, 6, 8, 10, 12, 14, 16, 18, 20)
+
+// pass in a lambda
+val c = nums.filter(_ % 2 == 0)   // List(2, 4, 6, 8, 10)
+val d = nums.map(_ * 2)           // List(2, 4, 6, 8, 10, 12, 14, 16, 18, 20)
+
+// methods can be chained as needed
+val e = nums.filter(_ > 3)        // List(40, 50, 60)
+            .filter(_ < 7)
+            .map(_ * 10)
+```
+
+In addition to HOFs used all throughout the standard library, you can easily create your own HOFs. This is shown in the Reference documentation.
+
+<!--
+### Functions are objects
+- add this section?
+- TODO: show the types of functions?
+    - see my 'val function vs def method page'
+-->
+
+
+## Extension methods
+
+<!-- TODO: i may not be showing the latest syntax -->
+*Extension methods* let you add new methods to closed classes. For instance, if you want to add a `hello` method to the `String` class, just create an extension method:
+
+<!-- TODO: scalafiddle -->
+```scala
+def (s: String) hello: String = s"Hello, ${s.capitalize}"
```

Comment:
The syntax has changed again, IIRC it should look like the following:

```suggestion
extension (s: String)
  def hello: String = s"Hello, ${s.capitalize}"
```

Reply:
Thanks. @b-studios caught this one, and I think I’m going to do something like this here:

```scala
extension (s: String):
  def hello: String = s"Hello, ${s.capitalize}"
  def aloha: String = s"Aloha, ${s.capitalize}"

"world".hello    // "Hello, World"
"friend".aloha   // "Aloha, Friend"
```